### PR TITLE
Spike/run tests in container

### PR DESCRIPTION
**Pull request by bwmcbride**
Sets up a container with the software needed to run tests + support scripts.

One minor mod to one of the tests to insert a delay after inserting characters into a form field, without which the test fails intermittently.

### CODE REVIEW

**Comment by bwmcbride** [comment]
I may not passed through all the relevant environment variables to the container when running a test.  See line 31 and following of bin/dkr-test.  Let me know if I need to add any more, or just add them.

File: bin/dkr-test [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+  docker run -it --rm \
+  --privileged \
+  --workdir=\"/home/tester\" \
+  --volume=\"$PWD/features:/home/tester/features:ro\" \
+  --volume=\"$PWD/specs:/home/tester/specs:ro\" \
+  --volume=\"$PWD/bin:/home/tester/bin:ro\" \
+  --volume=\"$PWD/lib:/home/tester/lib:ro\" \
+  --volume=/private/tmp/.X11-unix:/tmp.X11-unix \
+  -e IN_CONTAINER=1 \
+  -e DISPLAY=\"$ip:0\" \
```
Comment (andrew-pickin-epi) [inline bin/dkr-test:32]: This won't work in all cases (eg. where execution host is not display host. 
If DISPLAY is required pass it through as $DISPLAY

File: bin/dkr-test [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+docker_cmd=" \
+  docker run -it --rm \
+  --privileged \
+  --workdir=\"/home/tester\" \
+  --volume=\"$PWD/features:/home/tester/features:ro\" \
+  --volume=\"$PWD/specs:/home/tester/specs:ro\" \
+  --volume=\"$PWD/bin:/home/tester/bin:ro\" \
+  --volume=\"$PWD/lib:/home/tester/lib:ro\" \
+  --volume=/private/tmp/.X11-unix:/tmp.X11-unix \
+  -e IN_CONTAINER=1 \
```
Comment (andrew-pickin-epi) [inline bin/dkr-test:31]: I can't see this used anywhere

File: docker/Dockerfile [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+RUN   set -e ; \
+      apt-get update ; \
+      apt install -y firefox ;\
+      apt install -y ruby ; \
+      gem install bundler  ; \
+      apt install -y build-essential patch ruby-dev zlib1g-dev liblzma-dev ; \
+      apt install -y openjdk-8-jdk ; \
+      apt install -y wget ; \
+      apt install -y curl ; \
+      adduser --home /home/tester --shell /bin/bash --gecos "" --disabled-password tester
```
Comment (andrew-pickin-epi) [inline docker/Dockerfile:16]: apt-get install in one go.
&& \ is more conventional.

File: readme.md [hunk 4]
```diff
@@ -25,6 +25,47 @@ This runs:
 * legacy integration tests written in cucumber
 * revised integration tests for the updated UKHPI app, running under MiniTest
 
+## Running the tests in a Docker container
```
Comment (andrew-pickin-epi) [inline readme.md:28]: The ssh tunnel is horrid. Correct the security groups rather than kludging a hole through.

File: bin/dkr-build-test-container [hunk 5]
```diff
@@ -0,0 +1,10 @@
+#!/bin/bash
+
+set -e
+
+SCRIPT_DIR=$( dirname "$0" )
+source "$SCRIPT_DIR/dkr-config"
+
+echo docker image rm "${CONTAINER_NAME}:${CONTAINER_VERSION}" || true
```
Comment (andrew-pickin-epi) [inline bin/dkr-build-test-container:8]: It's not the Container name/version, but the image.

File: bin/dkr-build-test-container [hunk 6]
```diff
@@ -0,0 +1,10 @@
+#!/bin/bash
+
+set -e
+
+SCRIPT_DIR=$( dirname "$0" )
+source "$SCRIPT_DIR/dkr-config"
+
+echo docker image rm "${CONTAINER_NAME}:${CONTAINER_VERSION}" || true 
+docker build -f docker/Dockerfile  \
+       -t "${CONTAINER_NAME}:${CONTAINER_VERSION}" .
```
Comment (andrew-pickin-epi) [inline bin/dkr-build-test-container:10]: What are the content of the image supposed to be?
I'm finding hard to work out what the image should contain and what mouted?

File: bin/dkr-build-test-container (hunk 5 above)
Comment (andrew-pickin-epi) [inline bin/dkr-build-test-container:8]: Why is this line required?

File: bin/dkr-pull-test-container [hunk 7]
```diff
@@ -0,0 +1,12 @@
...
+
+# tag and push test container to ECR repo
+#
+
+set -e
+
+SCRIPT_DIR=$( dirname "$0" )
+source "$SCRIPT_DIR/dkr-config"
+
+docker pull "018852084843.dkr.ecr.eu-west-1.amazonaws.com/${CONTAINER_NAME}:${CONTAINER_VERSION}"
```
Comment (andrew-pickin-epi) [inline bin/dkr-pull-test-container:11]: Why are you pulling from the registry and then tagging.
This appears to be the opposite to convention.

File: bin/dkr-build-test-container [hunk 8]
```diff
@@ -0,0 +1,10 @@
+#!/bin/bash
```
[the remark this answers is not in the document]
Comment (bwmcbride) [reply bin/dkr-build-test-container:1]: I thought about using make.  Decided against on grounds:

1) the repo already uses rake instead of make.  Could have 2 make commands
2) I read make only supports sh level functionality and I wanted bash features
3) make's design centre is around automated update of things show source files change - which is not what is needed here.

Not a die in the ditch sorta thing.

If we have a policy of using make for things (which we don't at present) happy to change - and this could be the first to be updated.

File: bin/dkr-push-test-container [hunk 9]
```diff
@@ -0,0 +1,12 @@
+#!/bin/bash
```
[the remark this answers is not in the document]
Comment (bwmcbride) [reply bin/dkr-push-test-container:1]: This part of the general make decision.  Propose close this as a separate comment.

File: bin/dkr-test [hunk 10]
```diff
@@ -0,0 +1,55 @@
...
+
+SCRIPT_DIR=$( dirname "$0" )
+source "$SCRIPT_DIR/dkr-config"
+
+# build the docker command to run the tests in the container
+
+TEST_CMD="${TEST_CMD:-bash bin/test} "
+
+# this works on a linux or a mac: subsititue something based on ipconfig for windows
+ip=$(ifconfig en0 | grep inet | awk '$1=="inet" {print $2}')
```
[the remark this answers is not in the document]
Comment (bwmcbride) [reply bin/dkr-test:19]: Its not clear if this is an observation or a request for change and if a request for change, what to.

Is there something that is more standard that your propose using instead.

File: bin/dkr-test [hunk 11]
```diff
@@ -0,0 +1,55 @@
+#!/bin/bash
```
[the remark this answers is not in the document]
Comment (bwmcbride) [reply bin/dkr-test:1]: That violates the idea behind the container, as it is in my head at present.

The function of the container to collect together the right versions of the software needed to run the tests.  The test scripts and the tests themselves are deliberately outside the container to make them easy to edit, update and fiddle with, especially when test go wrong.

Propose reject this comment.

File: readme.md (hunk 4 above)
Comment (bwmcbride) [reply readme.md:28]: You may care to think that :)

Firstly, that's a value judgement, not a substantive comment of a problem.

Secondly, updating the security access controls is beyond the scope of the work for this pull request.  I hear what you say - but I think that updating the network, security controls etc should be subject of separate ticket.  Maybe we could have chat about alternatives and put such a ticket together?

Bear in mind, this was designed so that I could fix problems form a service station part way up the M6 at a time when I didn't have a working vpn to court lodge.

File: bin/dkr-build-test-container (hunk 6 above)
Comment (bwmcbride) [reply bin/dkr-build-test-container:10]: Principal is that container defines a software environment - most else is mounted.

File: bin/dkr-login [hunk 12]
```diff
@@ -0,0 +1,15 @@
+#!/bin/bash
```
[the remark this answers is not in the document]
Comment (bwmcbride) [reply bin/dkr-login:1]: Didn't know about that.  Ta.

File: bin/dkr-pull-test-container (hunk 7 above)
Comment (bwmcbride) [reply bin/dkr-pull-test-container:11]: It is set up so that you can build a container locally and use it without using the registry at all.  If you pull from the registry it sets a local tag so the scripts work.

Is there are a better way?

**Comment by bwmcbride** [comment]
@andrew-pickin-epi did you also have a comment about using $DISPLAY?  I thought I saw an email about it, but don't see the matching comment.

**Comment by ijdickinson** [comment]
> @andrew-pickin-epi did you also have a comment about using $DISPLAY? I thought I saw an email about it, but don't see the matching comment.

@bwmcbride See https://github.com/epimorphics/hmlr-apps-feature-tests/pull/9#discussion_r408788143

File: bin/dkr-test (hunk 1 above)
Comment (bwmcbride) [reply bin/dkr-test:32]: Missed this earlier.   Passing through $DISPLAY didn't work on a MAC when I tried it (permission problems and symlinks as I recall) - otherwise I would not have needed the fun and games to pick out the IP address.

I take your point about a remote X-server and will have another look.

**Comment by andrew-pickin-epi** [comment]
> > @andrew-pickin-epi did you also have a comment about using $DISPLAY? I thought I saw an email about it, but don't see the matching comment.
> 
> @bwmcbride See [#9 (comment)](https://github.com/epimorphics/hmlr-apps-feature-tests/pull/9#discussion_r408788143)

You can't try and work out what Display should be set to. eg my display is currently localhost:13.0.
Passing DISPLAY through unset is all you can hope to do.

File: bin/dkr-pull-test-container (hunk 7 above)
Comment (andrew-pickin-epi) [reply bin/dkr-pull-test-container:11]: Well you can but it's not the usual workflow.

File: bin/dkr-build-test-container (hunk 8 above)
[the remark this answers is not in the document]
Comment (andrew-pickin-epi) [reply bin/dkr-build-test-container:1]: It's not the dependency functionality that's significant. It pulls all you operations together in once place so as not to have a myriad of scripts and need to jump around the directory tree. 
As for the shell functions I  don't see any that are needed. Compare with my pr the makefile paired with dockerfile make it much easier to follow

File: bin/dkr-test (hunk 10 above)
[the remark this answers is not in the document]
Comment (andrew-pickin-epi) [reply bin/dkr-test:19]: if ifconfig isn't installed this will fail. even so setting display like this will only work in a limited number of cases

File: bin/dkr-test (hunk 11 above)
[the remark this answers is not in the document]
Comment (andrew-pickin-epi) [reply bin/dkr-test:1]: Still true, entrypoint would just be a different wrapper to cucumber. or rewrite bin/test to be neutral.

File: bin/dkr-login (hunk 12 above)
[the remark this answers is not in the document]
Comment (bwmcbride) [reply bin/dkr-login]: done

File: bin/dkr-build-test-container (hunk 8 above)
[the remark this answers is not in the document]
Comment (bwmcbride) [reply bin/dkr-build-test-container]: I have added the makefile.  It still uses scripts.

File: bin/dkr-push-test-container (hunk 9 above)
[the remark this answers is not in the document]
Comment (bwmcbride) [reply bin/dkr-push-test-container]: done

File: bin/dkr-test (hunk 10 above)
[the remark this answers is not in the document]
Comment (bwmcbride) [reply bin/dkr-test]: Yes it relies on ifconfig.  Its extensible to support windows but I haven't done that.

DISPLAY is now passed through unmodified if it points to remote server and is mapped to the local host if it doesn't so it works on a mac.

File: bin/dkr-build-test-container (hunk 5 above)
Comment (bwmcbride) [reply bin/dkr-build-test-container]: changed

Comment (bwmcbride) [reply bin/dkr-build-test-container]: An attempt to clean up

**Comment by bwmcbride** [comment]
I'm going to close this pull request and submit a new one based on a cleaned up merge of the latest changes in master.
